File: src/page/anim/IconBox.java
package page.anim;

import common.CommonStatic;
import common.CommonStatic.BCAuxAssets;
import common.system.fake.FakeGraphics;
import common.system.fake.FakeImage;
import page.view.ViewBox;
import utilpc.PP;

import java.awt.*;
import java.awt.event.KeyEvent;
import java.awt.event.MouseEvent;
import java.awt.image.BufferedImage;

import static page.anim.IconBox.IBConf.*;

public interface IconBox extends ViewBox {

    class IBConf {

        public static final int[] line = new int[4];
        public static int mode = 0, type = 0, glow = 0;

    }

    class IBCtrl extends ViewBox.Controller {

        protected boolean ctrl, drag;

        @Override
        public synchronized void mouseDragged(MouseEvent e) {
            if (!drag) {
                super.mouseDragged(e);
                return;
            }
            Point t = e.getPoint();
            line[ctrl ? 2 : 0] += t.x - p.x;
            line[ctrl ? 3 : 1] += t.y - p.y;
            p = t;
        }

        @Override
        public synchronized void mousePressed(MouseEvent e) {
            super.mousePressed(e);
            if (cont.isBlank() && !new PP(p).out(line, 1, -5))
                drag = true;
        }

        @Override
        public synchronized void mouseReleased(MouseEvent e) {
            drag = false;
            super.mouseReleased(e);
        }

        public void postdraw(FakeGraphics gra) {
            if (cont.isBlank()) {
                BCAuxAssets aux = CommonStatic.getBCAssets();
                int t = mode == 0 ? type == 1 ? 1 : 0 : 3;
                FakeImage bimg = aux.ico[mode][t].getImg();
                int bw = bimg.getWidth();
                int bh = bimg.getHeight();
                double r = Math.min(1.0 * line[2] / bw, 1.0 * line[3] / bh);
                gra.setColor(FakeGraphics.BLACK);
                gra.drawRect(line[0] - 1, line[1] - 1, line[2] + 1, line[3] + 1);
                if (glow == 1) {
                    gra.setComposite(FakeGraphics.BLEND, 256, 1);
                    bimg = aux.ico[0][4].getImg();
                    gra.drawImage(bimg, line[0], line[1], (int) (bw * r), (int) (bh * r));
                    gra.setComposite(FakeGraphics.DEF, 0, 0);
                }
                if (mode == 0 && type > 1) {
                    bimg = aux.ico[0][5].getImg();
                    gra.drawImage(bimg, line[0], line[1], (int) (bw * r), (int) (bh * r));
                } else {
                    bimg = aux.ico[mode][t].getImg();
                    gra.drawImage(bimg, line[0], line[1], (int) (bw * r), (int) (bh * r));
                }

            }
        }

        public void predraw(FakeGraphics gra) {
            if (cont.isBlank()) {
                BCAuxAssets aux = CommonStatic.getBCAssets();
                if (mode == 0 && type > 1 || mode == 1) {
                    FakeImage bimg = aux.ico[mode][type].getImg();
                    int bw = bimg.getWidth();
                    int bh = bimg.getHeight();
                    double r = Math.min(1.0 * line[2] / bw, 1.0 * line[3] / bh);
                    gra.drawImage(bimg, line[0], line[1], bw * r, bh * r);
                    if (glow == 1) {
                        gra.setComposite(FakeGraphics.BLEND, 256, -1);
                        bimg = aux.ico[0][4].getImg();
                        gra.drawImage(bimg, line[0], line[1], bw * r, bh * r);
                        gra.setComposite(FakeGraphics.DEF, 0, 0);
                    }
                }
            }
        }

        protected synchronized void keyPressed(KeyEvent ke) {
            if (ke.getKeyCode() == KeyEvent.VK_CONTROL)
                ctrl = true;
        }

        protected synchronized void keyReleased(KeyEvent ke) {
            if (ke.getKeyCode() == KeyEvent.VK_CONTROL)
                ctrl = false;
        }

    }

    void changeType();

    BufferedImage getClip();

    @Override
    IBCtrl getCtrl();

    default void keyPressed(KeyEvent ke) {
        getCtrl().keyPressed(ke);
    }

    default void keyReleased(KeyEvent ke) {
        getCtrl().keyReleased(ke);
    }

    void setBlank(boolean selected);

}

File: src/page/anim/IconBox.kt
package page.anim

import common.CommonStatic
import common.CommonStatic.BCAuxAssets
import common.system.fake.FakeGraphics
import common.system.fake.FakeImage
import page.view.ViewBox
import page.view.ViewBox.Controller
import utilpc.PP
import java.awt.event.KeyEvent
import java.awt.event.MouseEvent
import java.awt.image.BufferedImage

interface IconBox : ViewBox {
    object IBConf {
        val line = IntArray(4)
        var mode = 0
        var type = 0
        var glow = 0
    }

    class IBCtrl : Controller() {
        protected var ctrl = false
        protected var drag = false
        @Synchronized
        override fun mouseDragged(e: MouseEvent) {
            if (!drag) {
                super.mouseDragged(e)
                return
            }
            val t = e.point
            IBConf.line[if (ctrl) 2 else 0] += t.x - p.x
            IBConf.line[if (ctrl) 3 else 1] += t.y - p.y
            p = t
        }

        @Synchronized
        override fun mousePressed(e: MouseEvent) {
            super.mousePressed(e)
            if (cont.isBlank() && !PP(p).out(IBConf.line, 1.0, -5.0)) drag = true
        }

        @Synchronized
        override fun mouseReleased(e: MouseEvent?) {
            drag = false
            super.mouseReleased(e)
        }

        fun postdraw(gra: FakeGraphics) {
            if (cont.isBlank()) {
                val aux: BCAuxAssets = CommonStatic.getBCAssets()
                val t = if (IBConf.mode == 0) if (IBConf.type == 1) 1 else 0 else 3
                var bimg: FakeImage = aux.ico.get(IBConf.mode).get(t).getImg()
                val bw: Int = bimg.getWidth()
                val bh: Int = bimg.getHeight()
                val r = Math.min(1.0 * IBConf.line[2] / bw, 1.0 * IBConf.line[3] / bh)
                gra.setColor(FakeGraphics.Companion.BLACK)
                gra.drawRect(IBConf.line[0] - 1, IBConf.line[1] - 1, IBConf.line[2] + 1, IBConf.line[3] + 1)
                if (IBConf.glow == 1) {
                    gra.setComposite(FakeGraphics.Companion.BLEND, 256, 1)
                    bimg = aux.ico.get(0).get(4).getImg()
                    gra.drawImage(bimg, IBConf.line[0], IBConf.line[1], (bw * r) as Int.toDouble(), (bh * r) as Int.toDouble())
                    gra.setComposite(FakeGraphics.Companion.DEF, 0, 0)
                }
                if (IBConf.mode == 0 && IBConf.type > 1) {
                    bimg = aux.ico.get(0).get(5).getImg()
                    gra.drawImage(bimg, IBConf.line[0], IBConf.line[1], (bw * r) as Int.toDouble(), (bh * r) as Int.toDouble())
                } else {
                    bimg = aux.ico.get(IBConf.mode).get(t).getImg()
                    gra.drawImage(bimg, IBConf.line[0], IBConf.line[1], (bw * r) as Int.toDouble(), (bh * r) as Int.toDouble())
                }
            }
        }

        fun predraw(gra: FakeGraphics) {
            if (cont.isBlank()) {
                val aux: BCAuxAssets = CommonStatic.getBCAssets()
                if (IBConf.mode == 0 && IBConf.type > 1 || IBConf.mode == 1) {
                    var bimg: FakeImage = aux.ico.get(IBConf.mode).get(IBConf.type).getImg()
                    val bw: Int = bimg.getWidth()
                    val bh: Int = bimg.getHeight()
                    val r = Math.min(1.0 * IBConf.line[2] / bw, 1.0 * IBConf.line[3] / bh)
                    gra.drawImage(bimg, IBConf.line[0], IBConf.line[1], bw * r, bh * r)
                    if (IBConf.glow == 1) {
                        gra.setComposite(FakeGraphics.Companion.BLEND, 256, -1)
                        bimg = aux.ico.get(0).get(4).getImg()
                        gra.drawImage(bimg, IBConf.line[0], IBConf.line[1], bw * r, bh * r)
                        gra.setComposite(FakeGraphics.Companion.DEF, 0, 0)
                    }
                }
            }
        }

        @Synchronized
        fun keyPressed(ke: KeyEvent) {
            if (ke.keyCode == KeyEvent.VK_CONTROL) ctrl = true
        }

        @Synchronized
        fun keyReleased(ke: KeyEvent) {
            if (ke.keyCode == KeyEvent.VK_CONTROL) ctrl = false
        }
    }

    fun changeType()
    val clip: BufferedImage?
    val ctrl: IBCtrl
    fun keyPressed(ke: KeyEvent) {
        ctrl.keyPressed(ke)
    }

    fun keyReleased(ke: KeyEvent) {
        ctrl.keyReleased(ke)
    }

    fun setBlank(selected: Boolean)
}
